config/v1/types_infrastructure.go

```diff
@@ -505,9 +505,81 @@ type BareMetalPlatformStatus struct {
 	NodeDNSIP string `json:"nodeDNSIP,omitempty"`
 }
 
+// BGPPeer describes a BGP peer.
+type BGPPeer struct {
+	// asn is the Autonomous System Number of the peer.
+	//
+	// +optional
+	ASN string `json:"asn,omitempty"`
+
+	// ip is the IP address of the peer, as reachable from the cluster.
+	// It may be either IPv4 or IPv6.
+	//
+	// +kubebuilder:validation:Required
+	// + ---
+	// + {IPv4 or IPv6} validation in crd.yaml-patch
+	IP string `json:"ip"`
+
+	// password for BGP authentication against the peer.
+	//
+	// +optional
+	Password string `json:"password,omitempty"`
+}
+
+// OpenStackBGPSpeaker describes the BGP speaker configuration for a given
+// OpenStack subnet.
+type OpenStackBGPSpeaker struct {
+	// subnetID is the subnet which this BGP configuration applies to.
+	//
+	// +kubebuilder:validation:Required
+	SubnetID string `json:"subnetID"`
+
+	// asn specifies the Autonomous System number to be used by the BGP
+	// speaker. If this field is not set, the cluster will assign an
+	// arbitrary number.
+	//
+	// +optional
+	ASN string `json:"asn"`
+
+	// peers is a list of all BGP peers of the speaker of the subnet.
+	//
+	// +kubebuilder:validation:MinItems:=1
+	Peers []BGPPeer `json:"peers"`
+}
+
+type OpenStackBGPConfiguration struct {
+	// speakers is a list of OpenStack BGP speaker configurations. We require a
+	// speaker configuration for every subnet where we want to peer.
+	// The list must contain at least one item.
+	//
+	// +kubebuilder:validation:MinItems:=1
+	// +listType=map
+	// +listMapKey=subnetID
+	Speakers []OpenStackBGPSpeaker `json:"speakers"`
+}
+
 // OpenStackPlatformSpec holds the desired state of the OpenStack infrastructure provider.
 // This only includes fields that can be modified in the cluster.
-type OpenStackPlatformSpec struct{}
+type OpenStackPlatformSpec struct {
```

Comment:
if this information about how to write and modify a BGP API s supposed to function for other platforms at some future date (as indicated in the enhancement), please open another PR showing us what that would look like.  This doesn't immediately look so much like infrastructure versus networking configuration of some kind.  Especially if the information collected is actually generically used by other platforms.